SDK docs (#69)
cloudquery/docs: docs/developers/sdk/overview.md

# SDK Overview

> CloudQuery SDK enables building CloudQuery providers which allows CloudQuery's users to extract/load/transform existing and popular service providers as well as custom in-house solutions into a SQL or graph database.

Reviewer: Do we still enable loading into a graph?

Reviewer 2: No it should be removed for now, in future versions we might support a graph DB.

## Key Concepts

- **Resources** are defined as [schema.Table](https://github.com/cloudquery/cq-provider-sdk/blob/main/provider/schema/table.go) structs, which list available fields for a cloud resource along with metadata and table configuration.

> - **Resolvers** define how to fetch resources from the source (the cloud provider). [Utility resolvers](https://github.com/cloudquery/cq-provider-sdk/blob/main/provider/schema/resolvers.go) take this a step further and provide ready-made solutions on converting data from one data structure to another, ready to be saved in tables.

Reviewer: Should we expand in these docs or are we relying on the comments for each util function?

Author: It will be expanded.

# Overview

> CloudQuery SDK enables building CloudQuery providers which allows CloudQuery's users to extract/load/transform existing and popular service providers as well as custom in-house solutions into a SQL or graph database.

Reviewer 2: nit: I suggest to make E L T in the extract transform load bold

It was developed to allow easy creation for providers, and reduce boilerplate code when developing them.

The core idea of the SDK is to allow developing a provider in a straightforward manner, while allowing the user free control in fetching the data to be inserted.

## Key Concepts

> - **Provider** structs are the main core of the SDK that requires the implementor to create it list all available resources, configuration of it's client and it's config. The provider structs implements the CQProvider Interface allowing the user to just implement his resources and configure function.

Reviewer 2: ```suggestion
- **Provider** structs are the core component of the SDK that require the implementor to only set a list all available resources, configuration of it's client and it's config. The provider structs implements the CQProvider Interface allowing the user to just implement his resources and configure function.
```

- **Table** is the main building block in the SDK provider schema, these tables are passed to the Provider to define what resources the provider supports. [Tables](https://github.com/cloudquery/cq-provider-sdk/blob/main/provider/schema/table.go) define their columns, relations (which are also tables). Each table has a resolver function that is called by the SDK with the client that was configured early by the user implementation.
- **Resources** are wrappers around a single data that is fetched by table resolvers and passed to the SDK. Resources hold the original item and all resolved values that will be inserted into the database.  Terminologically speaking, a single Resource represents a row in the table.

> - **Resolvers** are functions to fetch resources from the source (the cloud provider). [Utility resolvers](https://github.com/cloudquery/cq-provider-sdk/blob/main/provider/schema/resolvers.go) take this a step further and provide ready-made solutions on converting data from one data structure to another, ready to be saved in tables.

Reviewer 2: I think we might want to distinct between TableResolver and ColumnResolver and PostResourceResolver?

Author: Done
